### PR TITLE
Bind expression variables appearing in a goto case statement.

### LINKED ISSUE #40714: C#: Typing goto case (ValueTuple<,>) in switch statement crashes Visual Studio
**Version Used**: 
VS2019 16.4.2
.NET Core 3.1

**Steps to Reproduce**:
1. Use the following code
```c#
static void Example(object a, object b)
{
    switch ((a, b))
    {
        case (string str, int[] arr) _:

            break;
        case (string str, decimal[] arr) _:
            break;
    }
}
```
2. Copy/paste ``goto case (string str, decimal[] arr)`` before the first ``break;``

**Expected Behavior**:
The goto case statement might not be valid, but Visual Studio should keep running.

**Actual Behavior**:
VS2019 16.4.2 crashes.

**Exception info from Windows event log**:
```
Application: devenv.exe
Framework Version: v4.0.30319
Description: The application requested process termination through System.Environment.FailFast(string message).
Message: System.InvalidOperationException: This program location is thought to be unreachable.
   at Microsoft.CodeAnalysis.CSharp.Binder.BindDeconstructionVariable(TypeWithAnnotations declTypeWithAnnotations, SingleVariableDesignationSyntax designation, CSharpSyntaxNode syntax, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindDeclarationVariablesForErrorRecovery(TypeWithAnnotations declTypeWithAnnotations, VariableDesignationSyntax node, CSharpSyntaxNode syntax, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindDeclarationExpressionAsError(DeclarationExpressionSyntax node, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindExpressionInternal(ExpressionSyntax node, DiagnosticBag diagnostics, Boolean invoked, Boolean indexed)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindExpression(ExpressionSyntax node, DiagnosticBag diagnostics, Boolean invoked, Boolean indexed)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindValue(ExpressionSyntax node, DiagnosticBag diagnostics, BindValueKind valueKind)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindTupleExpression(TupleExpressionSyntax node, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindExpressionInternal(ExpressionSyntax node, DiagnosticBag diagnostics, Boolean invoked, Boolean indexed)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindExpression(ExpressionSyntax node, DiagnosticBag diagnostics, Boolean invoked, Boolean indexed)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindNamespaceOrTypeOrExpression(ExpressionSyntax node, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.CSharpSemanticModel.Bind(Binder binder, CSharpSyntaxNode node, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.MethodBodySemanticModel.Bind(Binder binder, CSharpSyntaxNode node, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.MemberSemanticModel.GetBoundNodes(CSharpSyntaxNode node)
   at Microsoft.CodeAnalysis.CSharp.MemberSemanticModel.GetLowerBoundNode(CSharpSyntaxNode node)
   at Microsoft.CodeAnalysis.CSharp.MemberSemanticModel.GetTypeOfTupleLiteral(TupleExpressionSyntax declaratorSyntax)
   at Microsoft.CodeAnalysis.CSharp.MemberSemanticModel.GetDeclaredSymbol(TupleExpressionSyntax declaratorSyntax, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.SyntaxTreeSemanticModel.GetDeclaredSymbol(TupleExpressionSyntax declaratorSyntax, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CSharpSemanticModel.GetDeclaredSymbolCore(SyntaxNode node, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CSharpSemanticModel.GetDeclaredSymbolsCore(SyntaxNode declaration, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Diagnostics.SuppressMessageAttributeState.IsDiagnosticSuppressed(String id, Location location, Func`3 getSemanticModel, SuppressMessageInfo& info)
   at Microsoft.CodeAnalysis.Diagnostics.SuppressMessageAttributeState.ApplySourceSuppressions(Diagnostic diagnostic, Func`3 getSemanticModel, ISymbol symbolOpt)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.FilterDiagnosticsSuppressedInSource(ImmutableArray`1 diagnostics, Compilation compilation, SuppressMessageAttributeState suppressMessageState, Func`3 getSemanticModel)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.FilterDiagnosticsSuppressedInSourceOrByAnalyzers(ImmutableArray`1 diagnostics, Compilation compilation)
   at Microsoft.CodeAnalysis.Diagnostics.AnalysisResultBuilder.ApplySuppressionsAndStoreAnalysisResult(AnalysisScope analysisScope, AnalyzerDriver driver, Compilation compilation, Func`2 getAnalyzerActionCounts, Boolean fullAnalysisResultForAnalyzersInScope)
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.<ComputeAnalyzerDiagnosticsCoreAsync>d__64.MoveNext()
Stack:
   at System.Environment.FailFast(System.String, System.Exception)
   at Microsoft.CodeAnalysis.FailFast.OnFatalException(System.Exception)
   at Microsoft.CodeAnalysis.FatalError.Report(System.Exception, System.Action`1<System.Exception>)
   at Microsoft.CodeAnalysis.FatalError.ReportUnlessCanceled(System.Exception)
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers+<ComputeAnalyzerDiagnosticsCoreAsync>d__64.MoveNext()
   at Microsoft.CodeAnalysis.CSharp.Binder.BindDeconstructionVariable(Microsoft.CodeAnalysis.CSharp.Symbols.TypeWithAnnotations, Microsoft.CodeAnalysis.CSharp.Syntax.SingleVariableDesignationSyntax, Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode, Microsoft.CodeAnalysis.DiagnosticBag)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindDeclarationVariablesForErrorRecovery(Microsoft.CodeAnalysis.CSharp.Symbols.TypeWithAnnotations, Microsoft.CodeAnalysis.CSharp.Syntax.VariableDesignationSyntax, Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode, Microsoft.CodeAnalysis.DiagnosticBag)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindDeclarationExpressionAsError(Microsoft.CodeAnalysis.CSharp.Syntax.DeclarationExpressionSyntax, Microsoft.CodeAnalysis.DiagnosticBag)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindExpressionInternal(Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax, Microsoft.CodeAnalysis.DiagnosticBag, Boolean, Boolean)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindExpression(Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax, Microsoft.CodeAnalysis.DiagnosticBag, Boolean, Boolean)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindValue(Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax, Microsoft.CodeAnalysis.DiagnosticBag, BindValueKind)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindTupleExpression(Microsoft.CodeAnalysis.CSharp.Syntax.TupleExpressionSyntax, Microsoft.CodeAnalysis.DiagnosticBag)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindExpressionInternal(Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax, Microsoft.CodeAnalysis.DiagnosticBag, Boolean, Boolean)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindExpression(Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax, Microsoft.CodeAnalysis.DiagnosticBag, Boolean, Boolean)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindNamespaceOrTypeOrExpression(Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax, Microsoft.CodeAnalysis.DiagnosticBag)
   at Microsoft.CodeAnalysis.CSharp.CSharpSemanticModel.Bind(Microsoft.CodeAnalysis.CSharp.Binder, Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode, Microsoft.CodeAnalysis.DiagnosticBag)
   at Microsoft.CodeAnalysis.CSharp.MethodBodySemanticModel.Bind(Microsoft.CodeAnalysis.CSharp.Binder, Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode, Microsoft.CodeAnalysis.DiagnosticBag)
   at Microsoft.CodeAnalysis.CSharp.MemberSemanticModel.GetBoundNodes(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode)
   at Microsoft.CodeAnalysis.CSharp.MemberSemanticModel.GetLowerBoundNode(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode)
   at Microsoft.CodeAnalysis.CSharp.MemberSemanticModel.GetTypeOfTupleLiteral(Microsoft.CodeAnalysis.CSharp.Syntax.TupleExpressionSyntax)
   at Microsoft.CodeAnalysis.CSharp.MemberSemanticModel.GetDeclaredSymbol(Microsoft.CodeAnalysis.CSharp.Syntax.TupleExpressionSyntax, System.Threading.CancellationToken)
   at Microsoft.CodeAnalysis.CSharp.SyntaxTreeSemanticModel.GetDeclaredSymbol(Microsoft.CodeAnalysis.CSharp.Syntax.TupleExpressionSyntax, System.Threading.CancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CSharpSemanticModel.GetDeclaredSymbolCore(Microsoft.CodeAnalysis.SyntaxNode, System.Threading.CancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CSharpSemanticModel.GetDeclaredSymbolsCore(Microsoft.CodeAnalysis.SyntaxNode, System.Threading.CancellationToken)
   at Microsoft.CodeAnalysis.Diagnostics.SuppressMessageAttributeState.IsDiagnosticSuppressed(System.String, Microsoft.CodeAnalysis.Location, System.Func`3<Microsoft.CodeAnalysis.Compilation,Microsoft.CodeAnalysis.SyntaxTree,Microsoft.CodeAnalysis.SemanticModel>, Microsoft.CodeAnalysis.Diagnostics.SuppressMessageInfo ByRef)
   at Microsoft.CodeAnalysis.Diagnostics.SuppressMessageAttributeState.ApplySourceSuppressions(Microsoft.CodeAnalysis.Diagnostic, System.Func`3<Microsoft.CodeAnalysis.Compilation,Microsoft.CodeAnalysis.SyntaxTree,Microsoft.CodeAnalysis.SemanticModel>, Microsoft.CodeAnalysis.ISymbol)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.FilterDiagnosticsSuppressedInSource(System.Collections.Immutable.ImmutableArray`1<Microsoft.CodeAnalysis.Diagnostic>, Microsoft.CodeAnalysis.Compilation, Microsoft.CodeAnalysis.Diagnostics.SuppressMessageAttributeState, System.Func`3<Microsoft.CodeAnalysis.Compilation,Microsoft.CodeAnalysis.SyntaxTree,Microsoft.CodeAnalysis.SemanticModel>)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.FilterDiagnosticsSuppressedInSourceOrByAnalyzers(System.Collections.Immutable.ImmutableArray`1<Microsoft.CodeAnalysis.Diagnostic>, Microsoft.CodeAnalysis.Compilation)
   at Microsoft.CodeAnalysis.Diagnostics.AnalysisResultBuilder.ApplySuppressionsAndStoreAnalysisResult(Microsoft.CodeAnalysis.Diagnostics.AnalysisScope, Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver, Microsoft.CodeAnalysis.Compilation, System.Func`2<Microsoft.CodeAnalysis.Diagnostics.DiagnosticAnalyzer,Microsoft.CodeAnalysis.Diagnostics.Telemetry.AnalyzerActionCounts>, Boolean)
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers+<ComputeAnalyzerDiagnosticsCoreAsync>d__64.MoveNext()
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers+<ComputeAnalyzerDiagnosticsCoreAsync>d__64, Microsoft.CodeAnalysis, Version=3.4.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35]](<ComputeAnalyzerDiagnosticsCoreAsync>d__64 ByRef)
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.ComputeAnalyzerDiagnosticsCoreAsync(Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver, Microsoft.CodeAnalysis.Diagnostics.AsyncQueue`1<Microsoft.CodeAnalysis.Diagnostics.CompilationEvent>, Microsoft.CodeAnalysis.Diagnostics.AnalysisScope, System.Threading.CancellationToken)
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers+<>c__DisplayClass57_1+<<ComputeAnalyzerDiagnosticsAsync>b__1>d.MoveNext()
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers+<>c__DisplayClass57_1+<<ComputeAnalyzerDiagnosticsAsync>b__1>d, Microsoft.CodeAnalysis, Version=3.4.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35]](<<ComputeAnalyzerDiagnosticsAsync>b__1>d ByRef)
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers+<>c__DisplayClass57_1.<ComputeAnalyzerDiagnosticsAsync>b__1()
   at System.Threading.Tasks.Task`1[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
   at System.Threading.Tasks.Task.ExecutionContextCallback(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(System.Threading.Tasks.Task ByRef)
   at System.Threading.Tasks.Task.ExecuteEntry(Boolean)
   at System.Threading.Tasks.Task.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()
```

### PR DESCRIPTION
**Pull request by gafter**
Fixes #40714

### CODE REVIEW

**Comment by gafter** [comment]
@dotnet/roslyn-compiler May I please have a couple of reviews of this small bug fix?

File: src/Compilers/CSharp/Portable/Binder/LocalScopeBinder.cs [hunk 1]
```diff
@@ -247,6 +247,7 @@ protected ImmutableArray<LocalSymbol> BuildLocals(SyntaxList<StatementSyntax> st
                     case SyntaxKind.YieldReturnStatement:
                     case SyntaxKind.ReturnStatement:
                     case SyntaxKind.ThrowStatement:
+                    case SyntaxKind.GotoCaseStatement:
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/LocalScopeBinder.cs]: >SyntaxKind.GotoCaseStatement [](start = 25, length = 28)

Should we also handle regular GotoStatement?

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs [hunk 2]
```diff
@@ -1771,5 +1771,92 @@ static void M(int i)
                 Diagnostic(ErrorCode.ERR_AmbigUDConv, "new A()").WithArguments("A.implicit operator B(A)", "B.implicit operator B(A)", "A", "B").WithLocation(16, 43)
                 );
         }
+
+        [WorkItem(40714, "https://github.com/dotnet/roslyn/issues/40714")]
+        [Fact]
+        public void BadGotoCase_01()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs]: >BadGotoCase_01 [](start = 20, length = 14)

I think we should also test SemanticModel behavior: GetDeclaredSymbol, Get SymbolInfo, etc. The call stack in #40714 indicates that IDE requests information about declarations.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs]: Should probably test IOperation as well.

---
In reply to: [379658895](https://github.com/dotnet/roslyn/pull/41709#discussion_r379658895) [](ancestors = 379658895)

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1)

File: src/Compilers/CSharp/Portable/Binder/LocalScopeBinder.cs (hunk 1 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Binder/LocalScopeBinder.cs]: The parser cannot produce a node for a regular goto statement that has this issue.

**Comment by gafter** [comment]
@AlekseyTs I think I've responded to your comments.  Do you have any others?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs [hunk 3]
```diff
@@ -783,6 +785,9 @@ protected override TypeWithAnnotations InferTypeOfVarVariable(DiagnosticBag diag
                         var armBinder = (SwitchExpressionArmBinder)_nodeBinder;
                         armBinder.BindSwitchExpressionArm(arm, diagnostics);
                         break;
+                    case SyntaxKind.GotoCaseStatement:
+                        _nodeBinder.BindStatement((GotoStatementSyntax)_nodeToBind, diagnostics);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs]: >_nodeBinder.BindStatement((GotoStatementSyntax)_nodeToBind, diagnostics); [](start = 24, length = 73)

Was this breaking a test?

Comment (gafter) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs]: Yes, the new one. It also broke bootstrap builds but only in some non-Windows configuration, but I was not able to easily diagnose why.

**Pull request merged**